packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+     * property's key, `false` otherwise
+     */
+    function indexSignatureHasMatchingKey(
+      indexParameterType: ts.Type,
+      dynamicProperties: DynamicProperties,
+      remainingProperties: RemainingProperties,
+    ) {
+      const indexParameterTypeParts =
+        tsutils.unionTypeParts(indexParameterType);
+
+      const indexParameterHasNumberType = indexParameterTypeParts.some(
+        tsutils.isIntrinsicNumberType,
+      );
+
+      const indexParameterHasStringType = indexParameterTypeParts.some(
+        tsutils.isIntrinsicStringType,
+      );
+
+      // check remaining properties for a match
+      for (const key of remainingProperties.keys()) {
+        if (typeof key === 'number' && indexParameterHasNumberType) {
+          return true;
+        }
+
+        if (typeof key === 'string' && indexParameterHasStringType) {
+          return true;
+        }
+      }
+
+      // check dynamic properties for a match
+      for (const destructure of dynamicProperties) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.key,
+        );
+
+        for (const type of tsutils.unionTypeParts(destructure.type)) {
+          if (checker.isTypeAssignableTo(type, indexParameterType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * @returns `true` if an index signature's value type matches an existing
+     * property's value type, `false` otherwise
+     */
+    function indexSignatureHasMatchingValue(
+      remainingProperties: RemainingProperties,
+      indexSignature: TSESTree.TSIndexSignature,
+    ): boolean {
+      // `{ used }: { [i: `_${string}`] }`
+      if (!indexSignature.typeAnnotation) {
+        return true;
+      }
+
+      const indexType = services.getTypeAtLocation(
+        indexSignature.typeAnnotation.typeAnnotation,
+      );
+
+      for (const destructure of remainingProperties.values()) {
+        destructure.type ??= services.getTypeAtLocation(
+          destructure.property.value,
+        );
+
+        if (isTypeMatchingIndexSignatureType(destructure.type, indexType)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    /**
+     * Compares an index-signature value-type with another type. If `isNoUncheckedIndexedAccess`
+     * is enabled, will filter out `undefined` from the index-signature's type.
+     */
+    function isTypeMatchingIndexSignatureType(
+      indexSignatureType: ts.Type,
+      comparedType: ts.Type,
+    ): boolean {
+      if (isNoUncheckedIndexedAccess) {
+        const indexSignatureParts = tsutils
+          .unionTypeParts(indexSignatureType)
+          .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+        const comparedTypeParts = tsutils.unionTypeParts(comparedType);
+
+        if (indexSignatureParts.length !== comparedTypeParts.length) {
+          return false;
```

Comment:
[Testing] Uncovered line, as reported by codecov. Either this is reachable and should have a test covering it, or it's not and either a refactor or type assertion should be used to eliminate the dead code.